src/main/java/no/fint/betaling/service/FileService.java

```diff
@@ -0,0 +1,178 @@
+
+    @Value("${fint.betaling.dnd.file-types:application/vnd.openxmlformats-officedocument.spreadsheetml.sheet,application/vnd.ms-excel}")
+    private String[] fileTypes;
+    @Value("${fint.betaling.dnd.VIS-ID:VIS-ID}")
+    private String visId;
+
+    private final FileRepository fileRepository;
+
+    public FileService(FileRepository fileRepository) {
+        this.fileRepository = fileRepository;
+    }
+
+    public CustomerGroup getCustomers(String schoolId, ArrayList<String> customerNumbers) {
+        SkoleResource school = getSchool(schoolId);
+        CustomerGroup customerGroup = createCustomerGroup(school);
+        List<Customer> customerList = customerGroup.getCustomers().stream().filter(customer -> studentMatchStudentNumber(customer, customerNumbers)).collect(Collectors.toList());
+        customerGroup.setCustomers(customerList);
+        return customerGroup;
+    }
+
+    private boolean studentMatchStudentNumber(Customer customer, ArrayList<String> customerNumbers) {
+        return customerNumbers.contains(customer.getId());
+    }
+
+    private CustomerGroup createCustomerGroup(SkoleResource school) {
+        CustomerGroup customerGroup = new CustomerGroup();
+        customerGroup.setName(school.getNavn());
+        customerGroup.setDescription(null);
+        customerGroup.setCustomers(getCustomersForGroup(school.getElevforhold()));
+        return customerGroup;
+    }
+
+    private SkoleResource getSchool(String schoolId) {
+        return fileRepository.getSchools().values().stream()
+                .filter(hasOrganisationNumber(schoolId))
+                .findFirst()
+                .orElseThrow(() -> new SchoolNotFoundException(String.format("x-school-org-id: %s", schoolId)));
+    }
+
+    private Predicate<SkoleResource> hasOrganisationNumber(String schoolId) {
+        return s -> Optional.ofNullable(s.getOrganisasjonsnummer())
+                .map(Identifikator::getIdentifikatorverdi)
+                .map(schoolId::equals)
+                .orElse(false);
+    }
+
+    private List<Customer> getCustomersForGroup(List<Link> studentRelationLinks) {
+        if (studentRelationLinks == null) {
+            return Collections.emptyList();
+        }
+        Map<Link, ElevforholdResource> studentRelations = fileRepository.getStudentRelations();
+        Map<Link, PersonResource> students = fileRepository.getStudents();
+
+        return studentRelationLinks.stream()
+                .map(studentRelations::get)
+                .filter(Objects::nonNull)
+                .map(this::getStudentLink)
+                .filter(Objects::nonNull)
+                .map(students::get)
+                .filter(Objects::nonNull)
+                .map(CustomerFactory::toCustomer)
+                .distinct()
+                .collect(Collectors.toList());
+    }
+
+    private Link getStudentLink(ElevforholdResource resource) {
+        return resource.getElev().stream().findFirst().orElse(null);
```

Comment:
Return `Stream<Link>` instead, and use `.flatMap()` instead of `.map().filter(Objects::nonNull)`.

Reply:
I think you need to show my what you mean here.

Reply:
If `getStudentLink` returns `Stream<Link>`, you invoke this using `.flatMap(this::getStudentLink)` instead of `.map()`.

The subsequent actions of the stream would then be invoked for every Link, but would also work for the case where no link can be found. 

Then flatMap would get an empty stream, and simply stop.

Reply:
An attempt has been made, please have a look :)

Reply:
its deleted